src/System.IO.FileSystem/ref/System.IO.FileSystem.cs

```diff
@@ -251,8 +251,8 @@ namespace System.IO.Enumeration
     public ref struct FileSystemEntry
     {
         public ReadOnlySpan<char> Directory { get { throw null; } }
-        public string RootDirectory { get { throw null; } }
-        public string OriginalRootDirectory { get { throw null; } }
+        public ReadOnlySpan<char> RootDirectory { get { throw null; } }
+        public ReadOnlySpan<char> OriginalRootDirectory { get { throw null; } }
```

Comment:
I assume that this PR is as short as it is because we're storing strings into these properties and now it's still compiling because of the implicit cast?

The upside of this change is that we could return spans in the future.  The downside of this change is that in order to get these values now as strings, the developer will need to allocate, even though under the covers we already have the string we could have been giving them.

We're sure this is the right thing to do?  Are we going to be able to eliminate the strings currently being allocated?

Reply:
Usually you don't need to string these. They never change in the context of a enumeration, so you don't need to check identity either.

The plan is to have something that will unwrap strings, but even if we don't get that I could add a method in the future that would be optimized for our internal implementation. If we ship this way I'll be unable to avoid removing allocations.  `RootDirectory` could easily be removed once we've exposed normalization into a buffer.

Reply:
> The plan is to have something that will unwrap strings

You mean something that would let you do this?
```C#
string str = ...;
ReadOnlySpan<char> span = sr;
string orig = Unwrap(span);
```
without allocating?

Reply:
> You mean something that would let you do this?

Yes, something like that. I spoke with @KrzysztofCwalina about this- intent is to do this. It has come up a few times in the review of this particular feature and I can envision this optimization being quite useful.

Reply:
> about this- intent is to do this

@jkotas, really?

Reply:
> @jkotas, really?

I have said multiple times that I do not think having a reverse conversion of (fast) Span back to String or array is doable, with reasonable performance characteristic.

Reply:
Right, that's what I thought.

Reply:
My understanding was some sort of static helper / extension method. In the context of this change it isn't critical- if such a thing is really needed we can add ToOriginalRootDirectory() or something similar. I don't think it will be as you have all of the context to avoid allocating for these two properties.

It is also possible that we might want to create constructor overloads that allow passing in a directory as a span. I'm still mulling on that one as I'm not entirely convinced it is super helpful. The design goal here is allocations per move next. While you could chain enumerators together slightly more efficiently that way you're still allocating the enumerator...

Reply:
yeah, my impression was that it would work only for slow span.

Reply:
> yeah, my impression was that it would work only for slow span.

Good to know. Even if we can't unwrap here (even slowly), I wouldn't change back to string. It is easily possible for developers to work around a potential extra allocation here if we have ReadOnlySpan<>. It isn't easily possible if we change the internal implementation to not use string and ship returning string.

Reply:
Don't lose sight of the fact that the other two "string" properties on the struct are already ReadOnlySpan<char> (Directory and FileName). One of them _must_ be. The other was changed from string to ReadOnlySpan<char> based on the initial reviews, explicitly so we retained the flexibility to further optimize the implementation.

Reply:
> Even if we can't unwrap here (even slowly)

The problem is that the "slowly" would likely be so costly as to completely make it useless, and worse, actually cause perf problems.

> I wouldn't change back to string

Ok.